### ISSUE TITLE
webproxycache update causing some https sites to fail

### ISSUE THREAD

**Issue by jedrington** [RedHatEMEA/demobuilder#148]
It looks like submodule commit RedHatEMEA/webproxycache@fc9133c546068f7f2145b5d1bac68209d37fc243 is causing some https sites to fail. This is most noticeable when using a fresh clone of demobuilder and making an initial rhel-server-7 image build. It is possible that others have not run into this issue yet if they are working on higher layers above the OS or still using an older version of the submodule.

Notably, not all SSL/TLS sites are impacted. For instance, communication to subscription.rhn.redhat.com:443/subscription is still working since it was able to register the system and yum install the nfs-utils package. The first case where I found the failure was when vm-functions attempts to download the default vagrant authorized_keys file.

Here are selected lines of build output:
```
Tue Mar  1 19:02:08 2016: rhel-server-7:vagrant-libvirt: + su vagrant -c 'mkdir -m 0700 /home/vagrant/.ssh; https_proxy=http://192.168.122.1:1025/ curl -sLo /home/vagrant/.ssh/authorized_keys https://github.com/mitchellh/vagrant/raw/master/keys/vagrant.pub; chmod 0600 /home/vagrant/.ssh/authorized_keys'
Tue Mar  1 19:02:08 2016: rhel-server-7:vagrant-libvirt: chmod: cannot access ‘/home/vagrant/.ssh/authorized_keys’: No such file or directory
```

Here are the steps to reproduce the issue:
````
git clone --recursive https://github.com/RedHatEMEA/demobuilder.git
cd demobuilder
utils/init.sh
vim config.yml
ln -s ~/Downloads/iso/rhel-server-7.2-x86_64-dvd.iso isos/rhel-server-7.2-x86_64-dvd.iso [modify as appropriate]
./build.py rhel-server-7:vagrant-libvirt
```

I believe I've isolated the issue to the above mentioned commit. Attempting a build with the prior commit is successful, as tested here:
```
cd contrib/webproxycache
git reset --hard b51ec8d36e3a02f4ac13da94228c48c9f25d56f0
cd ../..
git commit -a
./build.py rhel-server-7:vagrant-libvirt
```
It is also possible to make this issue occur at a place where you can ssh into the instance. Manually running the curl command results in the following error message. As expected, the curl command works correctly when omitting the https_proxy variable and bypassing the webproxycache.
```
[vagrant@localhost root]$ https_proxy=http://192.168.122.1:1025/ curl -vL https://github.com/mitchellh/vagrant/raw/master/keys/vagrant.pub
* About to connect() to proxy 192.168.122.1 port 1025 (#0)
*   Trying 192.168.122.1...
* Connected to 192.168.122.1 (192.168.122.1) port 1025 (#0)
* Establish HTTP proxy tunnel to github.com:443
> CONNECT github.com:443 HTTP/1.1
> Host: github.com:443
> User-Agent: curl/7.29.0
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 200 Connection established
< 
* Proxy replied OK to CONNECT request
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* NSS error -5938 (PR_END_OF_FILE_ERROR)
* Encountered end of file
* Closing connection 0
curl: (35) Encountered end of file
```

I hope this issue is detailed enough. I'd be glad to provide more information if needed.

**Comment by jim-minter** [comment]
Hi, thanks for the exemplary bug report.

Please can you confirm these issues are happening with webproxycache HEAD (a9f1dd1fa7763e9bbc3cbced85b2616fb7488184) ?

When it goes wrong, please can you e-mail me the output of:

```
https_proxy=http://192.168.122.1:1025/ curl -L --trace - https://github.com/mitchellh/vagrant/raw/master/keys/vagrant.pub
```

As well as the content of the resultant `log` file in the demobuilder root directory?

As you're debugging, depending on whether the issue turns out to be with cached or uncached downloads, you may need to delete the webproxycache.db file between successive runs as this is where the cache state is saved.

I'd be surprised if the commit you mention was the problem one, but the introduction of HEAD support into webproxycache around 27/1 was painful and I've found GitHub downloads are particularly brittle.

The main issue with GitHub downloads was because sometimes they are sent with an HTTP Content-Length field, and sometimes not, depending on the download server you hit.  The variance depending on the backend server can make it tricky to catch the error (which may mean your bisect isn't correct).  Nonetheless, I thought I'd got it all fixed and I haven't been seeing any issues on the latest HEAD (which is not to say there *isn't* a bug, of course!).

Thanks for the report, happy to look at this further if you can get a curl trace.

**Comment by jedrington** [comment]
Hi Jim, my pleasure!

> Please can you confirm these issues are happening with webproxycache HEAD (a9f1dd1fa7763e9bbc3cbced85b2616fb7488184)?

Yes.

> The variance depending on the backend server can make it tricky to catch the error (which may mean your bisect isn't correct). Nonetheless, I thought I'd got it all fixed and I haven't been seeing any issues on the latest HEAD (which is not to say there isn't a bug, of course!).

Understood, although I'm not sure that's what we're dealing with. Maybe the trace and log output will give you a better idea of the problem.

I originally discovered the issue when using a new recursive clone, which used webproxycache HEAD (a9f1dd1fa7763e9bbc3cbced85b2616fb7488184). After a bit of debugging, I found the last commit that worked properly for me was RedHatEMEA/webproxycache@b51ec8d36e3a02f4ac13da94228c48c9f25d56f0. I've tested these same three commits going on a dozen or two times now, and I've never once seen any variation in result.

This afternoon, I re-ran all the tests, while collecting the requested log files and observing the behavior. Perhaps we've actually stumbled onto two separate issues. Let me elaborate...

##### Testing Process
I started with clean recursive clone and performed a `./build.py rhel-server-7:vagrant-libvirt` multiple times (removing webproxycache.db each time) in the following order:

Test Description | webproxycache HEAD | Result | log error | Observation
-------------------- | ----------------------------- | --------- | ----------- | ---------------
1. Latest code (recursive) | RedHatEMEA/webproxycache@a9f1dd1fa7763e9bbc3cbced85b2616fb7488184 | fail | IOError: [Errno 2] No such file or directory: 'certs/ca.crt' | No certs directory created at first run
2. Rolled back to 'rework sslca' commit | RedHatEMEA/webproxycache@fc9133c546068f7f2145b5d1bac68209d37fc243 | fail | IOError: [Errno 2] No such file or directory: 'certs/ca.crt' | Same behavior as latest commit
3. Rolled back to 'missing flush' commit | RedHatEMEA/webproxycache@b51ec8d36e3a02f4ac13da94228c48c9f25d56f0 | pass | n/a | certs/ca.crt exists

The proxy command at this point is GET http://cacert/ HTTP/1.1.

With the success of step 3, I now have a working releases/rhel-server-7:vagrant-libvirt.box. In items 1 and 2 above, since the failure occurred during the kickstart %post of the rhel-server-7 layer build, we never got to point where a valid base build existed, and therefore no ability to ssh for debugging. With the success of step 3, if we remove releases/rhel-server-7:vagrant-libvirt.box and rerun scenario 1 or 2, we skip the rhel-server-7 build and when the error occurs while trying to curl the vagrant ssh key (with debug enabled) during the rhel-server-7:vagrant-libvirt layer build, I can then ssh in to run additional traces.

It was at this point I noticed a different error in the log file... which in retrospect makes perfect sense. certs/ca.crt now exists because of the third run, so we don't hit that issue again. But instead we hit another seemingly similar, but probably distinctly different, issue.

Test Description | webproxycache HEAD | Result | log error | Observation
-------------------- | ----------------------------- | --------- | ----------- | ---------------
4. Latest code (recursive) | RedHatEMEA/webproxycache@a9f1dd1fa7763e9bbc3cbced85b2616fb7488184 | fail | NameError: global name 'CA' is not defined | 
5. Rolled back to 'rework sslca' commit | RedHatEMEA/webproxycache@fc9133c546068f7f2145b5d1bac68209d37fc243 | fail | NameError: global name 'CA' is not defined |
6. Rolled back to 'missing flush' commit | RedHatEMEA/webproxycache@b51ec8d36e3a02f4ac13da94228c48c9f25d56f0 | pass | n/a | Works as expected.

The proxy command at this point is CONNECT https://github.com:443 HTTP/1.1. 

I'll also email you the log file and console / trace output separately. Thank you!

**Comment by jim-minter** [comment]
I think this was an idiotic initialisation bug on my part.  I think that d8e2339478a59769c69980c8ae1f39c19e5d8104 should fix it - please try that.

**Comment by jedrington** [comment]
Yes, it was resolved by d8e2339. Thank you!
